### PR TITLE
feat: App Update Module.

### PR DESCRIPTION
**Pull request by merlin-northern**
Changelog: Title
Ticket: MEN-6076
Signed-off-by: Peter Grzybowski <peter@northern.tech>

### CODE REVIEW

File: src/app [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    fi
+
+    $tar_decompress_cmd "$1"/images.tar.gz -C "$temp_dir"
+    $tar_decompress_cmd "$1"/manifests.tar.gz -C "$temp_dir"
+
+    for image_dir in "${temp_dir}/images/"*; do
+        url_new=$(cat "${image_dir}/url-new.txt")
+        url_current=$(cat "${image_dir}/url-current.txt")
+        sha_new=$(cat "${image_dir}/sums-new.txt")
+        sha_current=$(cat "${image_dir}/sums-current.txt")
```
Comment (lluiscampos) [inline src/app:77]: I would also sanity check all these variables

File: src/app [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+if test -d "$persistent_store"; then
+    if touch "$persistent_store"/.rw_test; then
+        rm -f "$persistent_store"/.rw_test
+    else
+        echo "ERROR: cant write to persistent_store in $persistent_store"
+        exit 1
+    fi
+fi
+
+delta_cmd="xdelta3 -d -s"
```
Comment (lluiscampos) [inline src/app:39]: Can we use long options for `-d` and `-s`?

File: src/app [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        mv -v "${persistent_store}/${application_name}/" "${persistent_store}/${application_name}-${rollback_id}"
+        $app_sub_module stop "${persistent_store}/${application_name}-${rollback_id}/manifests" "${application_name}"
+    fi
+    mkdir -p "${persistent_store}/${application_name}"
+    mv -v "$temp_dir/manifests" "${persistent_store}/${application_name}/"
+    $app_sub_module rollout "${persistent_store}/${application_name}/manifests" "${application_name}"
+    # now if $app_sub_module healthy then rm -Rf "${persistent_store}/${artifact_name}-${rollback_id}"
+    # and on rollback we just deploy "${persistent_store}/${artifact_name}-${rollback_id}"
+    # earlier ln -s "${persistent_store}/${artifact_name}-${rollback_id}" "${persistent_store}/${artifact_name}-rollback"
+    # and removing all else
```
Comment (lluiscampos) [inline src/app:106]: This I didn't fully understand at first glance. I need to come back to it :sweat_smile:

File: src/app [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    version=$(jq -r .version < "$1")
+    app_sub_module="${app_module_dir}/${orchestrator}"
+    if test ! -f "${app_sub_module}"; then
+        echo "ERROR: ${app_sub_module} not found. exiting."
+        return 1
+    fi
+    artifact_name=$(jq -r .artifact_provides.artifact_name < "$2")
+}
+
+decompress_artifact() {
```
Comment (lluiscampos) [inline src/app:55]: This function does more than "decompress" the artifact. Please rename to something more meaningful like `handle_artifact` (?).

File: src/app-modules/docker-compose [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#!/bin/sh
+
+set -x
+
+docker_compose_cmd=docker-compose
+docker_cmd=docker
+
+app_load() {
+    local url="$1"
```
Comment (lluiscampos) [inline src/app-modules/docker-compose:9]: Unused.

If the intention is to leave it for future reference, then add also `local application_name="$3"`. Otherwise remove :+1:

File: src/app-modules/docker-compose [hunk 6]
```diff
@@ -0,0 +1,51 @@
+#!/bin/sh
+
```
Comment (lluiscampos) [inline src/app-modules/docker-compose:2]: Missing license header

File: src/app-modules/docker-compose [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    local url="$1"
+    local output_file="$2"
+
+    $docker_cmd image save "$url" -o "$output_file"
+}
+
+app_rollout() {
+    local manifests_dir="$1"
+
+    (cd "$manifests_dir" && $docker_compose_cmd up -d > compose.log 2>&1 3>&1) &
```
Comment (lluiscampos) [inline src/app-modules/docker-compose:25]: Can you ELI5 what `3>&1` is doing here?

Comment (lluiscampos) [inline src/app-modules/docker-compose:25]: (here and below)

Instead of doing `cd "$manifests_dir" && ...`, could we just use `--project-directory` option?

File: src/app-modules/docker-compose [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+app_stop() {
+    local manifests_dir="$1"
+
+    (cd "$manifests_dir" && $docker_compose_cmd down)
+}
+
+case "$1" in
+    save)
```
Comment (lluiscampos) [inline src/app-modules/docker-compose:35]: The [API docs](https://github.com/mendersoftware/app-update-module/blob/master/docs/README-submodule-api.md) seems to indicate that these should be uppercase.

File: src/app-modules/docker-compose [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        app_load "$@"
+        ;;
+    rollout)
+        shift
+        app_rollout "$@"
+        ;;
+    stop)
+        shift
+        app_stop "$@"
+        ;;
```
Comment (lluiscampos) [inline src/app-modules/docker-compose:50]: For completeness, I would add the rest of the valid commands with a `true` command and then a `*)` case with an error. Or you don't want it to be strict for future expansions of the API?

File: src/app-modules/docker-compose [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+app_save() {
+    local url="$1"
+    local output_file="$2"
+
+    $docker_cmd image save "$url" -o "$output_file"
+}
+
+app_rollout() {
+    local manifests_dir="$1"
```
Comment (lluiscampos) [inline src/app-modules/docker-compose:23]: This is wrong, should be `$2` according to the API docs.

File: src/app-modules/docker-compose [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+    (cd "$manifests_dir" && $docker_compose_cmd up -d > compose.log 2>&1 3>&1) &
+}
+
+app_stop() {
+    local manifests_dir="$1"
```
Comment (lluiscampos) [inline src/app-modules/docker-compose:29]: Also here, according to the (amended) docs this should be `$2`.

File: docs/README-submodule-api.md [hunk 12]
```diff
@@ -191,12 +191,16 @@ Stops the application
...
+* `source_directory` -- directory containing the manifests. In case of k8s
+  it is a directory where we run apply, or docker-compose up in case of docker-compose.
 
 #### START
 Starts the application
 
 Parameters:
 * `application_name` -- application name
+* `source_directory` -- directory containing the manifests. In case of k8s
+  it is a directory where we run apply, or docker-compose up in case of docker-compose.
```
Comment (lluiscampos) [inline docs/README-submodule-api.md:204]: Don't you think we will need to add also `source_directory` to  rollback command?

**Review by lluiscampos — CHANGES_REQUESTED** [review]
Overall it looks very well structured and easy to read :+1: 

Quite a few comments. Some are minor suggestions so use your best judgement!

File: .gitlab-ci.yml [hunk 13]
```diff
@@ -13,5 +13,5 @@ include:
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply .gitlab-ci.yml:11]: fixed.

File: conf/mender-app.conf [hunk 14]
```diff
@@ -0,0 +1 @@
+persistent_store=/data/mender-app
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply conf/mender-app.conf:1]: fixed.

File: src/app [hunk 15]
```diff
@@ -0,0 +1,131 @@
+#!/bin/sh
+
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app:15]: added. shall I create a QA ticket for it? I have 0 capacity to do it now)

File: src/app [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+set -e
+
+temp_dir=$(mktemp -d "app.tmp.XXXXXXXX")
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app]: thank you, completely forgot about it. move to use it.

File: src/app (hunk 2 above)
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app:54]: do you honestly believe that I forgot about it? https://tracker.mender.io/browse/MEN-6084 says: "So we need to include xdelta3 as a dependency.". I added a note to [the debs one](https://tracker.mender.io/browse/MEN-6196) though.

File: src/app [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    application_name=$(jq -r .application_name < "$1")
+    platform=$(jq -r .platform < "$1")
+    orchestrator=$(jq -r .orchestrator < "$1")
+    version=$(jq -r .version < "$1")
+    app_sub_module="${app_module_dir}/${orchestrator}"
+    if test ! -f "${app_sub_module}"; then
+        echo "ERROR: ${app_sub_module} not found. exiting."
+        return 1
+    fi
+    artifact_name=$(jq -r .artifact_provides.artifact_name < "$2")
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app:75]: fixed. all are checked now.

File: src/app [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+config_file="/etc/mender/mender-app.conf"
+version=""
+orchestrator=""
+platform=""
+manifests_dir=""
+app_module_dir="/usr/share/mender/app-modules/v1"
+app_sub_module=""
+artifact_name=""
+persistent_store="/data/mender-app"
+application_name=""
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app]: since I changed the consts to upper case, and I believe it looks better now, I decided not to prefix further. is that ok?

File: src/app [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    test -f "$persistent_store"/.rw_test && rm -f "$persistent_store"/.rw_test
+}
+
+trap cleanup 1 2 3 6 15
+
+config_file="/etc/mender/mender-app.conf"
+version=""
+orchestrator=""
+platform=""
+manifests_dir=""
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app]: removed the manifests_dir as it was truly unused. the rest is filled, occasionally unused, but I would like to keep it, as those are vital informations, which maybe needed. I understand unused variables removal, I would keep them, as they represent what came with the artifact.

File: src/app [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+delta_cmd="xdelta3 -d -s"
+tar_decompress_cmd="tar -xzvf"
+
+parse_metadata() {
+    application_name=$(jq -r .application_name < "$1")
+    platform=$(jq -r .platform < "$1")
+    orchestrator=$(jq -r .orchestrator < "$1")
+    version=$(jq -r .version < "$1")
+    app_sub_module="${app_module_dir}/${orchestrator}"
+    if test ! -f "${app_sub_module}"; then
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app:63]: added. I also added the check if the file is readable.

File: src/app [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+decompress_artifact() {
+    local image_dir
+    local image
+    local url_new
+    local url_current
+    local sha_new
+    local sha_current
+    local current_image
+    local new_image
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app]: I moved to `image_new` and `image_current`. perhaps it will be ok, since it is only the output path where we save the binary data of the application components. I wanted to move from image, but it is not a bad name.

File: src/app [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    local rollback_id="last"
+
+    if test "$temp_dir" = ""; then
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app:114]: added the check (there is no way mktemp will return non-error and non-existing directory, but it is fine, it may have been deleted in the meantime).

File: src/app [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+        return 1
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app:121]: added.

File: src/app (hunk 1 above)
Comment (merlin-northern) [reply src/app:131]: done.

File: src/app (hunk 2 above)
Comment (merlin-northern) [reply src/app:54]: as far as I know there are no long options for `xdelta3` for -s and -d.

File: src/app (hunk 3 above)
Comment (merlin-northern) [reply src/app]: let's have a relaxing tea over a call (or a call over relaxing tea).

File: src/app (hunk 4 above)
Comment (merlin-northern) [reply src/app]: renamed to `handle_artifact`.

File: src/app-modules/docker-compose (hunk 5 above)
Comment (merlin-northern) [reply src/app-modules/docker-compose]: added the `application_name`.

File: src/app-modules/docker-compose (hunk 6 above)
Comment (merlin-northern) [reply src/app-modules/docker-compose:15]: fixed.

File: src/app-modules/docker-compose (hunk 7 above)
Comment (merlin-northern) [reply src/app-modules/docker-compose]: we need to detach the command from the terminal. the `disown` is not supported by `/bin/sh`, `setsid` is not always present (would be another dependency, you know), `nohup` can also be absent. that was the best I could come up with.

Comment (merlin-northern) [reply src/app-modules/docker-compose]: fixed.

File: src/app-modules/docker-compose (hunk 8 above)
Comment (merlin-northern) [reply src/app-modules/docker-compose]: fixed.

File: src/app-modules/docker-compose (hunk 9 above)
Comment (merlin-northern) [reply src/app-modules/docker-compose:67]: added.

File: src/app-modules/docker-compose (hunk 10 above)
Comment (merlin-northern) [reply src/app-modules/docker-compose]: fixed.

File: src/app-modules/docker-compose (hunk 11 above)
Comment (merlin-northern) [reply src/app-modules/docker-compose]: fixed. I also went through remaing API calls, and make them accept always application name as the first argument.

File: docs/README-submodule-api.md (hunk 12 above)
Comment (merlin-northern) [reply docs/README-submodule-api.md:206]: yes. good catch. added.

File: .gitlab-ci.yml [hunk 24]
```diff
@@ -3,15 +3,36 @@ stages:
...
+  tags:
+    - docker
+  image: docker:20.10.10-dind-alpine3.14
+  services:
+    - name: docker:20.10.10-dind-alpine3.14
+      alias: docker
+      command: ["--mtu=1440"] # https://gitlab.com/gitlab-com/gl-infra/production/-/issues/5590
+  before_script:
+    - apk --update --no-cache add bash curl aws-cli tree xz jq wget xdelta3
+    - curl -L https://github.com/docker/compose/releases/download/v2.15.1/docker-compose-linux-x86_64 -o /usr/bin/docker-compose
```
Comment (lluiscampos) [inline .gitlab-ci.yml:30]: Can we use image `tiangolo/docker-with-compose` instead, as we do in other places in CI?

Otherwise we risk getting the dependency out of date quickly and unnoticed.

File: src/app (hunk 15 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/app:15]: No problem, I'll handle it.

File: src/app (hunk 2 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/app:54]: > do you honestly believe that I forgot about it? (...)

Better noisy :mega: than sorry :fire:

Comment (lluiscampos) [reply src/app:54]: Okey :see_no_evil:

File: src/app (hunk 20 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/app:63]: I think that executable is enough, actually. Why in theory do we need to have read rights?

Additionally, if is executable it means that is also a file, so I'd say that `test -f` is not necessary.

File: src/app (hunk 22 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/app:114]: Yes, `mktemp` guarantees the directory. But if you wanted to check something, check that it exists rather than that the variable is set.

This first check now can be removed.

File: src/app-modules/docker-compose [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+        true
+        ;;
+    ALIVE)
+        true
+        ;;
+    HEALTHY)
+        true
+        ;;
+    *)
+        echo "ERROR: unknown API call: $@"
```
Comment (lluiscampos) [inline src/app-modules/docker-compose:93]:
```suggestion
        echo "ERROR: unknown API call: $1"
```

File: tests/app/data/docker/manifests-delta/docker-compose.yml [hunk 26]
```diff
@@ -0,0 +1,23 @@
+version: '2.1'
```
Comment (lluiscampos) [inline tests/app/data/docker/manifests-delta/docker-compose.yml:1]: Once we are requiring 2.3 in [other parts of CI](https://github.com/mendersoftware/integration/pull/2314), I think is okey to modernize here too :+1:

File: tests/app/data/docker/manifests-delta/docker-compose.yml [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+        image: docker.io/library/memcached:1.6.19-alpine
+        networks:
+            - local_net
+        environment:
+            - POSTGRES_HOST_AUTH_METHOD=trust
+        entrypoint:
+            - /bin/sleep
+            - "512"
+networks:
+    local_net: null
```
Comment (lluiscampos) [inline tests/app/data/docker/manifests-delta/docker-compose.yml:23]: No new line at end of file

File: tests/app/scenarios/00_basic_ArtifactInstall.run.sh [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+sleep 4
+cat /data/mender-app/mapp64/manifests/*.yml
+cat /data/mender-app/mapp64/manifests/*.log
+grep -rniHF Pulling /data/mender-app/mapp64/manifests/*.log && false
+echo
+sleep 16
+docker ps --format "{{.Image}} {{.Command}}" | sort
+diff <(docker ps --format "{{.Image}} {{.Command}}" | sort) <(cat "${EXPECTED_CONTAINERS}")
+[[ -d "${DATA}"/output/ ]] || mkdir "${DATA}"/output/
+cp -a /tests/* "${DATA}"/output/
```
Comment (lluiscampos) [inline tests/app/scenarios/00_basic_ArtifactInstall.run.sh:55]: Why don't we do tests in Python like every other repository?

File: src/app (hunk 18 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/app]: Yes, that was also my preference. The prefix was an alternative.

File: src/app (hunk 19 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/app]: Yes, the comment was only about `manifests_dir`.

File: src/app (hunk 21 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/app]: It is not a bad name, it is only incomplete in my opinion. Because the other variables (url, sha) also refer to images (image url, image sha).

Is okey, though. Not a big deal.

File: src/app-modules/docker-compose (hunk 7 above)
Comment (lluiscampos) [reply src/app-modules/docker-compose]: Yes. And how creating a new file descriptor (`3`) and redirect there stdin (`&1`) detaches the command? (I just don't know)

File: src/app-modules/docker-compose (hunk 8 above)
Comment (lluiscampos) [reply src/app-modules/docker-compose]: But do we really want it in uppercase? I thought you would fix the docs instead :sweat_smile: 

It is good with me both ways, as long as doc and code say the same thing

File: src/app [hunk 29]
```diff
@@ -0,0 +1,199 @@
...
+    fi
+    if test ! -r "${app_sub_module}"; then
+        echo "ERROR: ${app_sub_module} not readable. exiting."
+        return 1
+    fi
+    artifact_name=$(jq -r .artifact_provides.artifact_name < "$2")
+
+    if test "${application_name}" = ""; then
+        echo "ERROR: application_name is required. exiting."
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply src/app:78]: >I think that executable is enough, actually. Why in theory do we need to have read rights?

 you cant execute something you cant read. since you wanted the exec flag check, then it made me add readable check as well.

File: .gitlab-ci.yml (hunk 24 above)
Comment (merlin-northern) [reply .gitlab-ci.yml:30]: I would have to check. I hope it is alpine based, otherwise lots of work.

File: tests/app/scenarios/00_basic_ArtifactInstall.run.sh (hunk 28 above)
Comment (merlin-northern) [reply tests/app/scenarios/00_basic_ArtifactInstall.run.sh:55]: we do not do tests in python in every repository. there is a new class of repos where we write quick tests in shell, for the ease of run and ease of use. are you frontally against those?

Comment (merlin-northern) [reply tests/app/scenarios/00_basic_ArtifactInstall.run.sh:55]: besides: this test emulates a device, are you sure it would be better in python?

File: .gitlab-ci.yml [hunk 30]
```diff
@@ -3,15 +3,36 @@ stages:
...
-    file: '.gitlab-ci-check-license.yml'
-  - project: 'Northern.tech/Mender/mendertesting'
-    file: '.gitlab-ci-github-status-updates.yml'
-  - project: 'Northern.tech/Mender/mendertesting'
-    file: '.gitlab-ci-check-shell-format.yml'
+    file:
+      - '.gitlab-ci-check-commits.yml'
+      - '.gitlab-ci-check-license.yml'
+      - '.gitlab-ci-github-status-updates.yml'
+      - '.gitlab-ci-check-shell-format.yml'
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply .gitlab-ci.yml:10]: > This should be on a separate commit.

ok I figured I will put just all test-related stuff in one commit. moved.

File: .gitlab-ci.yml [hunk 31]
```diff
@@ -3,15 +3,36 @@ stages:
...
 
 test:check-shell-formatting:
   before_script:
-    - SHELL_SCRIPTS=/dev/null
+    - SHELL_SCRIPTS="src/app src/app-modules/docker-compose tests/app/run.sh tests/app/scenarios/00_basic_ArtifactInstall.run.sh"
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply .gitlab-ci.yml:14]: ok, moved.

**Comment by merlin-northern** [comment]
>But do we really want it in uppercase? I thought you would fix the docs instead 😅

> It is good with me both ways, as long as doc and code say the same thing

I honestly have no idea, why the reply to comment option in github is not avaiable. I have to copy the comments as quotes. ok, so now it is all upper case, and consistent.

File: conf/mender-app.conf [hunk 32]
```diff
@@ -0,0 +1 @@
+PERSISTENT_STORE=/data/mender-app
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply conf/mender-app.conf]: ok now github is loosing messages, I aleardy answered to that one, the message is lost.

>No new line at the end of file. GitHub said so
so we have to fix it :) personally, I only listem to Lluís.

**Comment by merlin-northern** [comment]
>Yes. And how creating a new file descriptor (3) and redirect there stdin (&1) detaches the command? (I just don't know)

at one point there must have been another open fd, but now I re-tested, currently it is not needed, so I stand corrected. ramoved.

File: .gitlab-ci.yml (hunk 30 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply .gitlab-ci.yml:10]: Thank you.

File: .gitlab-ci.yml (hunk 31 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply .gitlab-ci.yml]: Thank you. Still a tiny bit off after the rebase from master but I stop here :)

[the remark this answers is not in the document]
Comment (lluiscampos) [reply .gitlab-ci.yml]: Thank you. Still a tiny bit off after the rebase from master but I stop here :)

File: .gitlab-ci.yml [hunk 33]
```diff
@@ -3,36 +3,35 @@ stages:
...
+      - '.gitlab-ci-github-status-updates.yml'
+      - '.gitlab-ci-check-shell-format.yml'
 
 test:check-shell-formatting:
   before_script:
-    - SHELL_SCRIPTS="gen/app-gen"
+    - SHELL_SCRIPTS="src/app src/app-modules/docker-compose gen/app-gen tests/app/run.sh tests/app/scenarios/00_basic_ArtifactInstall.run.sh"
 
-test:smoketests:gen:
+test:smoketests:app:
```
Comment (lluiscampos) [inline .gitlab-ci.yml:16]: This job is rather `test:smoketests:`

File: src/app-modules/docker-compose (hunk 25 above)
Comment (lluiscampos) [reply src/app-modules/docker-compose:93]: This went missing.

File: tests/app/data/docker/manifests-delta/docker-compose.yml [hunk 34]
```diff
@@ -0,0 +1,23 @@
+version: '2.3'
```
Comment (lluiscampos) [inline tests/app/data/docker/manifests-delta/docker-compose.yml:1]: `docker-compose-dev.yml` files are unused. Remove?

File: tests/app/scenarios/00_basic_ArtifactInstall.run.sh (hunk 28 above)
Comment (lluiscampos) [reply tests/app/scenarios/00_basic_ArtifactInstall.run.sh:55]: Okey, were are we then writing acceptance tests, will it be in [MEN-6082](https://tracker.mender.io/browse/MEN-6082)? Could we review the epic from QA point of view on a later call?

File: .gitlab-ci.yml (hunk 31 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply .gitlab-ci.yml]: Thank you. Still a tiny bit off after the rebase from master but I stop here :)

File: .gitlab-ci.yml (hunk 24 above)
Comment (lluiscampos) [reply .gitlab-ci.yml]: Worth for the sake of maintenance, IMO. Thanks.

File: src/app [hunk 35]
```diff
@@ -0,0 +1,199 @@
...
+    $tar_decompress_cmd "$1"/images.tar.gz -C "$temp_dir"
+    $tar_decompress_cmd "$1"/manifests.tar.gz -C "$temp_dir"
+
+    for image_dir in "${temp_dir}/images/"*; do
+        url_new=$(cat "${image_dir}/url-new.txt")
+        url_current=$(cat "${image_dir}/url-current.txt")
+        sha_new=$(cat "${image_dir}/sums-new.txt")
+        sha_current=$(cat "${image_dir}/sums-current.txt")
+        if test "$url_new" = ""; then
+            echo "ERROR: url_new cannot be empty"
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/app]: This needs to apply to all errors below (or to any) to be consistent. I will be **explicit** next time.

**Review by lluiscampos — APPROVED** [review]

File: tests/app/scenarios/00_basic_ArtifactInstall.run.sh (hunk 28 above)
Comment (merlin-northern) [reply tests/app/scenarios/00_basic_ArtifactInstall.run.sh:55]: > Okey. Were are we then writing acceptance tests, will it be in [MEN-6082](https://tracker.mender.io/browse/MEN-6082)? Could we review the epic from QA point of view on a later call?

yes; all the remaining test coverage will go with MEN-6080 that is the plan. and sure. 1125?

File: tests/app/data/docker/manifests-delta/docker-compose.yml (hunk 34 above)
Comment (merlin-northern) [reply tests/app/data/docker/manifests-delta/docker-compose.yml:1]: removed! (it was present in the artifact as a place holder and to show that you can have more than one file, unused here, thank you)

File: .gitlab-ci.yml (hunk 33 above)
Comment (merlin-northern) [reply .gitlab-ci.yml]: > This job is rather `test:smoketests:`

renamed!

File: src/app-modules/docker-compose (hunk 25 above)
Comment (merlin-northern) [reply src/app-modules/docker-compose:93]: > This went missing.

pardon me? I am sorry, I do not get it, it is in the code, you want me to do something here?

**Comment by merlin-northern** [comment]
> This went missing.

pardon me? I am sorry,  it is in the code, do you want me to do something with it?

**Comment by merlin-northern** [comment]
> Okey. Were are we then writing acceptance tests, will it be in [MEN-6082](https://tracker.mender.io/browse/MEN-6082)? Could we review the epic from QA point of view on a later call?

I have no idea what is going on, github looses comments, I am typing this one for the fourth time. of course I would like to chat 1125? and yes, in MEN-6080 as per design we are supposed to complete the coverage.
